File: frontend/src/api/supplier-invoices.ts
import { apiClient } from './client';
import type { SuccessResponse } from '../types/api';

export interface SupplierInvoiceItem {
  id: string;
  supplier_invoice_id: string;
  spo_item_id: string | null;
  grn_item_id: string | null;
  product_id: string;
  description: string;
  quantity: number;
  uom_id: string;
  unit_price: number;
  discount_percent: number;
  vat_rate: number;
  vat_amount: number;
  total_price: number;
  currency: string;
  match_status: string;
  variance_quantity: number;
  variance_price: number;
  variance_tax: number;
  variance_notes: string | null;
}

export interface SupplierInvoice {
  id: string;
  workspace_id: string;
  supplier_id: string;
  supplier_invoice_number: string;
  our_reference: string | null;
  invoice_date: string;
  due_date: string;
  currency: string;
  subtotal: number;
  discount_amount: number;
  vat_amount: number;
  total_amount: number;
  amount_paid: number;
  balance_due: number;
  status: 'RECEIVED' | 'PENDING_MATCHING' | 'MATCHED' | 'DISCREPANCY' | 'APPROVED' | 'PARTIALLY_PAID' | 'PAID' | 'CANCELLED';
  three_way_match_status: string;
  three_way_match_notes: string | null;
  document_url: string | null;
  ocr_job_id: string | null;
  ocr_extracted: boolean;
  primary_spo_id: string | null;
  received_at: string;
  matched_at: string | null;
  approved_at: string | null;
  paid_at: string | null;
  created_at: string;
  updated_at: string;
  items: SupplierInvoiceItem[];
}

export interface SupplierInvoiceItemCreate {
  spo_item_id?: string | null;
  grn_item_id?: string | null;
  product_id: string;
  description: string;
  quantity: number;
  uom_id: string;
  unit_price: number;
  discount_percent?: number;
  vat_rate?: number;
  vat_amount?: number;
  total_price?: number;
  currency: string;
}

export interface SupplierInvoiceCreate {
  supplier_id: string;
  supplier_invoice_number: string;
  our_reference?: string | null;
  invoice_date: string;
  due_date: string;
  currency: string;
  subtotal: number;
  discount_amount: number;
  vat_amount: number;
  total_amount: number;
  primary_spo_id?: string | null;
  items: SupplierInvoiceItemCreate[];
}

export const getSupplierInvoices = async (): Promise<SupplierInvoice[]> => {
  const response = await apiClient.get<SuccessResponse<SupplierInvoice[]>>('/api/v1/supplier-invoices');
  return response.data.data;
};

export const getSupplierInvoice = async (id: string): Promise<SupplierInvoice> => {
  const response = await apiClient.get<SuccessResponse<SupplierInvoice>>(`/api/v1/supplier-invoices/${id}`);
  return response.data.data;
};

export const createSupplierInvoice = async (data: SupplierInvoiceCreate): Promise<SupplierInvoice> => {
  const response = await apiClient.post<SuccessResponse<SupplierInvoice>>('/api/v1/supplier-invoices', data);
  return response.data.data;
};

export const submitMatching = async (id: string): Promise<SupplierInvoice> => {
  const response = await apiClient.post<SuccessResponse<SupplierInvoice>>(`/api/v1/supplier-invoices/${id}/submit-matching`);
  return response.data.data;
};

export const resolveDiscrepancy = async (id: string, notes: string): Promise<SupplierInvoice> => {
  const response = await apiClient.post<SuccessResponse<SupplierInvoice>>(`/api/v1/supplier-invoices/${id}/resolve-discrepancy`, { notes });
  return response.data.data;
};

export const approveInvoice = async (id: string): Promise<SupplierInvoice> => {
  const response = await apiClient.post<SuccessResponse<SupplierInvoice>>(`/api/v1/supplier-invoices/${id}/approve`);
  return response.data.data;
};

File: backend/app/schemas/supplier_invoices.py
from decimal import Decimal
from pydantic import BaseModel, ConfigDict, Field
from typing import List, Optional
from datetime import datetime
import uuid

from app.models.supplier_invoice import SupplierInvoiceStatus, MatchResult
from app.models.supplier_invoice_event import SupplierInvoiceEventType


class SupplierInvoiceItemBase(BaseModel):
    spo_item_id: Optional[uuid.UUID] = None
    grn_item_id: Optional[uuid.UUID] = None
    product_id: uuid.UUID
    description: str
    quantity: Decimal = Field(..., ge=0)
    uom_id: uuid.UUID
    unit_price: Decimal = Field(..., ge=0)
    discount_percent: Decimal = Field(default=Decimal("0.00"), ge=0)
    vat_rate: Decimal = Field(default=Decimal("0.00"), ge=0)
    vat_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    total_price: Decimal = Field(default=Decimal("0.00"), ge=0)
    currency: str = Field(max_length=3)


class SupplierInvoiceItemCreate(SupplierInvoiceItemBase):
    pass


class SupplierInvoiceItemResponse(SupplierInvoiceItemBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    supplier_invoice_id: uuid.UUID
    match_status: MatchResult
    variance_quantity: Decimal
    variance_price: Decimal
    variance_tax: Decimal
    variance_notes: Optional[str]


class SupplierInvoiceBase(BaseModel):
    supplier_id: uuid.UUID
    supplier_invoice_number: str
    our_reference: Optional[str] = None
    invoice_date: datetime
    due_date: datetime
    currency: str = Field(max_length=3)
    subtotal: Decimal = Field(default=Decimal("0.00"), ge=0)
    discount_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    vat_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    total_amount: Decimal = Field(default=Decimal("0.00"), ge=0)
    primary_spo_id: Optional[uuid.UUID] = None


class SupplierInvoiceCreate(SupplierInvoiceBase):
    items: List[SupplierInvoiceItemCreate]


class SupplierInvoiceResponse(SupplierInvoiceBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID
    amount_paid: Decimal
    balance_due: Decimal
    status: SupplierInvoiceStatus
    three_way_match_status: MatchResult
    three_way_match_notes: Optional[str]
    document_url: Optional[str]
    ocr_job_id: Optional[str]
    ocr_extracted: bool
    received_at: datetime
    matched_at: Optional[datetime]
    approved_at: Optional[datetime]
    paid_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    items: List[SupplierInvoiceItemResponse]


class SupplierInvoiceDiscrepancyResolution(BaseModel):
    notes: str


class SupplierInvoiceEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    supplier_invoice_id: uuid.UUID
    workspace_id: uuid.UUID
    event_type: SupplierInvoiceEventType
    previous_status: Optional[str]
    new_status: str
    actor_id: uuid.UUID
    metadata_log: dict
    created_at: datetime
